Controller fails when TFJob is complete

/kind bug

I tried to run Katib TFJob example: https://github.com/kubeflow/katib/blob/master/examples/v1beta1/tfjob-example.yaml with the new Training manager.

As I can see, my TFJob `CleanPodPolicy` is empty.
Once TFJob is complete and Katib controller deletes Succeeded TFJob, I am getting this error in the controller:

```
2021-08-14T00:21:26.792Z	DEBUG	controller-runtime.manager.events	Normal	{"object": {"kind":"TFJob","namespace":"kubeflow","name":"tfjob-example-j6b7ff7q","uid":"ad20a360-7468-4358-8657-66c73bd4f4a0","apiVersion":"kubeflow.org/v1","resourceVersion":"345415368"}, "reason": "TFJobSucceeded", "message": "TFJob kubeflow/tfjob-example-j6b7ff7q successfully completed."}
time="2021-08-14T00:21:26Z" level=info msg="Finished updating TFJobs Status \"tfjob-example-j6b7ff7q\" (19.43709ms)" job=kubeflow.tfjob-example-j6b7ff7q uid=ad20a360-7468-4358-8657-66c73bd4f4a0
time="2021-08-14T00:21:26Z" level=info msg="Reconciling for job tfjob-example-j6b7ff7q"
E0814 00:21:26.818156       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 353 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1642580, 0x2655f50)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1642580, 0x2655f50)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/kubeflow/common/pkg/controller.v1/common.(*JobController).DeletePodsAndServices(0xc0004ad200, 0xc00019b638, 0x17dce00, 0xc00019b520, 0xc0009d2f20, 0x2, 0x2, 0x1817f8b, 0x9)
	/go/pkg/mod/github.com/kubeflow/common@v0.3.4/pkg/controller.v1/common/job.go:27 +0x3c
github.com/kubeflow/common/pkg/controller.v1/common.(*JobController).ReconcileJobs(0xc0004ad200, 0x17dce00, 0xc00019b520, 0xc000624cc0, 0xc000cbcc60, 0x3, 0x3, 0xc000624cf0, 0xc000b72c60, 0xc000b72c80, ...)
	/go/pkg/mod/github.com/kubeflow/common@v0.3.4/pkg/controller.v1/common/job.go:161 +0x68f
github.com/kubeflow/tf-operator/pkg/controller.v1/tensorflow.(*TFJobReconciler).Reconcile(0xc0004ad200, 0x1a82060, 0xc000624c30, 0xc000994140, 0x8, 0xc00079f000, 0x16, 0xc000624c30, 0x40903b, 0xc000030000, ...)
	/workspace/pkg/controller.v1/tensorflow/tfjob_controller.go:153 +0x834
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000882c80, 0x1a81fa0, 0xc0008da640, 0x169fa20, 0xc000b72b20)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:263 +0x2f1
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000882c80, 0x1a81fa0, 0xc0008da640, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:235 +0x202
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1a81fa0, 0xc0008da640)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000907f50)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000d77f50, 0x1a45ba0, 0xc00092e270, 0xc0008da601, 0xc00055b6e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000907f50, 0x3b9aca00, 0x0, 0x1, 0xc00055b6e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1a81fa0, 0xc0008da640, 0xc000ac6f90, 0x3b9aca00, 0x0, 0x1)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1a81fa0, 0xc0008da640, 0xc000ac6f90, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:195 +0x4f6
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1425b3c]

goroutine 353 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1642580, 0x2655f50)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/kubeflow/common/pkg/controller.v1/common.(*JobController).DeletePodsAndServices(0xc0004ad200, 0xc00019b638, 0x17dce00, 0xc00019b520, 0xc0009d2f20, 0x2, 0x2, 0x1817f8b, 0x9)
	/go/pkg/mod/github.com/kubeflow/common@v0.3.4/pkg/controller.v1/common/job.go:27 +0x3c
github.com/kubeflow/common/pkg/controller.v1/common.(*JobController).ReconcileJobs(0xc0004ad200, 0x17dce00, 0xc00019b520, 0xc000624cc0, 0xc000cbcc60, 0x3, 0x3, 0xc000624cf0, 0xc000b72c60, 0xc000b72c80, ...)
	/go/pkg/mod/github.com/kubeflow/common@v0.3.4/pkg/controller.v1/common/job.go:161 +0x68f
github.com/kubeflow/tf-operator/pkg/controller.v1/tensorflow.(*TFJobReconciler).Reconcile(0xc0004ad200, 0x1a82060, 0xc000624c30, 0xc000994140, 0x8, 0xc00079f000, 0x16, 0xc000624c30, 0x40903b, 0xc000030000, ...)
	/workspace/pkg/controller.v1/tensorflow/tfjob_controller.go:153 +0x834
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000882c80, 0x1a81fa0, 0xc0008da640, 0x169fa20, 0xc000b72b20)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:263 +0x2f1
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000882c80, 0x1a81fa0, 0xc0008da640, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:235 +0x202
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1a81fa0, 0xc0008da640)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000907f50)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000d77f50, 0x1a45ba0, 0xc00092e270, 0xc0008da601, 0xc00055b6e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000907f50, 0x3b9aca00, 0x0, 0x1, 0xc00055b6e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1a81fa0, 0xc0008da640, 0xc000ac6f90, 0x3b9aca00, 0x0, 0x1)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1a81fa0, 0xc0008da640, 0xc000ac6f90, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.9/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:195 +0x4f6
```


cc @kubeflow/wg-training-leads

--- a/pkg/controller.v1/mxnet/mxjob_controller.go
+++ b/pkg/controller.v1/mxnet/mxjob_controller.go
@@ -33,7 +33,6 @@ import (
 	"github.com/kubeflow/common/pkg/controller.v1/expectation"
 	commonutil "github.com/kubeflow/common/pkg/util"
 	mxjobv1 "github.com/kubeflow/tf-operator/pkg/apis/mxnet/v1"
-	"github.com/kubeflow/tf-operator/pkg/client/clientset/versioned/scheme"
 	"github.com/kubeflow/tf-operator/pkg/common/util"
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
@@ -157,7 +156,7 @@ func (r *MXJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	}
 
 	// Set default priorities to mxnet job
-	scheme.Scheme.Default(mxjob)
+	r.Scheme.Default(mxjob)
 
 	// Convert PyTorch.Spec.PyTorchReplicasSpecs to  map[commonv1.ReplicaType]*commonv1.ReplicaSpec
 	replicas := map[commonv1.ReplicaType]*commonv1.ReplicaSpec{}
@@ -196,7 +195,7 @@ func (r *MXJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 	// using onOwnerCreateFunc is easier to set defaults
 	if err = c.Watch(&source.Kind{Type: &mxjobv1.MXJob{}}, &handler.EnqueueRequestForObject{},
-		predicate.Funcs{CreateFunc: onOwnerCreateFunc()},
+		predicate.Funcs{CreateFunc: r.onOwnerCreateFunc()},
 	); err != nil {
 		return err
 	}
@@ -448,17 +447,15 @@ func (r *MXJobReconciler) IsMasterRole(replicas map[commonv1.ReplicaType]*common
 }
 
 // onOwnerCreateFunc modify creation condition.
-func onOwnerCreateFunc() func(event.CreateEvent) bool {
+func (r *MXJobReconciler) onOwnerCreateFunc() func(event.CreateEvent) bool {
 	return func(e event.CreateEvent) bool {
 		mxjob, ok := e.Object.(*mxjobv1.MXJob)
 		if !ok {
 			return true
 		}
 
-		mxjobv1.SetDefaults_MXJob(mxjob)
-		// TODO: figure out why default funcs are not registered successfully.
 		// Use defaulters registered in scheme.
-		scheme.Scheme.Default(mxjob)
+		r.Scheme.Default(mxjob)
 		msg := fmt.Sprintf("xgboostJob %s is created.", e.Object.GetName())
 		logrus.Info(msg)
 

--- a/pkg/controller.v1/pytorch/pytorchjob_controller.go
+++ b/pkg/controller.v1/pytorch/pytorchjob_controller.go
@@ -38,7 +38,6 @@ import (
 	"github.com/kubeflow/common/pkg/controller.v1/control"
 	"github.com/kubeflow/common/pkg/controller.v1/expectation"
 	pytorchv1 "github.com/kubeflow/tf-operator/pkg/apis/pytorch/v1"
-	"github.com/kubeflow/tf-operator/pkg/client/clientset/versioned/scheme"
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
@@ -148,7 +147,7 @@ func (r *PyTorchJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 	}
 
 	// Set default priorities to pytorch job
-	scheme.Scheme.Default(pytorchjob)
+	r.Scheme.Default(pytorchjob)
 
 	// Construct RunPolicy based on PyTorchJob.Spec
 	runPolicy := &commonv1.RunPolicy{
@@ -181,7 +180,7 @@ func (r *PyTorchJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 	// using onOwnerCreateFunc is easier to set defaults
 	if err = c.Watch(&source.Kind{Type: &pytorchv1.PyTorchJob{}}, &handler.EnqueueRequestForObject{},
-		predicate.Funcs{CreateFunc: onOwnerCreateFunc()},
+		predicate.Funcs{CreateFunc: r.onOwnerCreateFunc()},
 	); err != nil {
 		return err
 	}
@@ -438,15 +437,13 @@ func (r *PyTorchJobReconciler) IsMasterRole(replicas map[commonv1.ReplicaType]*c
 }
 
 // onOwnerCreateFunc modify creation condition.
-func onOwnerCreateFunc() func(event.CreateEvent) bool {
+func (r *PyTorchJobReconciler) onOwnerCreateFunc() func(event.CreateEvent) bool {
 	return func(e event.CreateEvent) bool {
 		pytorchjob, ok := e.Object.(*pytorchv1.PyTorchJob)
 		if !ok {
 			return true
 		}
-		pytorchv1.SetDefaults_PyTorchJob(pytorchjob)
-		// TODO: figure out why default funcs are not registered successfully.
-		scheme.Scheme.Default(pytorchjob)
+		r.Scheme.Default(pytorchjob)
 		msg := fmt.Sprintf("PyTorchJob %s is created.", e.Object.GetName())
 		logrus.Info(msg)
 		if err := commonutil.UpdateJobConditions(&pytorchjob.Status, commonv1.JobCreated, "PyTorchJobCreated", msg); err != nil {

--- a/pkg/controller.v1/tensorflow/tfjob_controller.go
+++ b/pkg/controller.v1/tensorflow/tfjob_controller.go
@@ -59,7 +59,6 @@ import (
 
 	"github.com/go-logr/logr"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/client-go/kubernetes/scheme"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
@@ -147,7 +146,7 @@ func (r *TFJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	}
 
 	// Set default priorities to tfjob
-	scheme.Scheme.Default(tfjob)
+	r.Scheme.Default(tfjob)
 
 	// Use common to reconcile the job related pod and service
 	err = r.ReconcileJobs(tfjob, tfjob.Spec.TFReplicaSpecs, tfjob.Status, &tfjob.Spec.RunPolicy)
@@ -171,7 +170,7 @@ func (r *TFJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 	// using onOwnerCreateFunc is easier to set defaults
 	if err = c.Watch(&source.Kind{Type: &tfv1.TFJob{}}, &handler.EnqueueRequestForObject{},
-		predicate.Funcs{CreateFunc: onOwnerCreateFunc()},
+		predicate.Funcs{CreateFunc: r.onOwnerCreateFunc()},
 	); err != nil {
 		return err
 	}
@@ -835,16 +834,14 @@ func (r *TFJobReconciler) createNewPod(tfjob *tfv1.TFJob, rt, index string, spec
 }
 
 // onOwnerCreateFunc modify creation condition.
-func onOwnerCreateFunc() func(event.CreateEvent) bool {
+func (r *TFJobReconciler) onOwnerCreateFunc() func(event.CreateEvent) bool {
 	return func(e event.CreateEvent) bool {
 		tfJob, ok := e.Object.(*tensorflowv1.TFJob)
 		if !ok {
 			return true
 		}
 
-		tensorflowv1.SetDefaults_TFJob(tfJob)
-		// TODO: figure out why default funcs are not registered successfully.
-		scheme.Scheme.Default(tfJob)
+		r.Scheme.Default(tfJob)
 		msg := fmt.Sprintf("TFJob %s is created.", e.Object.GetName())
 		logrus.Info(msg)
 

--- a/pkg/controller.v1/xgboost/xgboostjob_controller.go
+++ b/pkg/controller.v1/xgboost/xgboostjob_controller.go
@@ -44,7 +44,6 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	kubeclientset "k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/log"
@@ -163,7 +162,7 @@ func (r *XGBoostJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 	}
 
 	// Set default priorities for xgboost job
-	scheme.Scheme.Default(xgboostjob)
+	r.Scheme.Default(xgboostjob)
 
 	// Use common to reconcile the job related pod and service
 	err = r.ReconcileJobs(xgboostjob, xgboostjob.Spec.XGBReplicaSpecs, xgboostjob.Status, &xgboostjob.Spec.RunPolicy)
@@ -187,7 +186,7 @@ func (r *XGBoostJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 	// using onOwnerCreateFunc is easier to set defaults
 	if err = c.Watch(&source.Kind{Type: &xgboostv1.XGBoostJob{}}, &handler.EnqueueRequestForObject{},
-		predicate.Funcs{CreateFunc: onOwnerCreateFunc()},
+		predicate.Funcs{CreateFunc: r.onOwnerCreateFunc()},
 	); err != nil {
 		return err
 	}
@@ -444,15 +443,13 @@ func (r *XGBoostJobReconciler) IsMasterRole(replicas map[commonv1.ReplicaType]*c
 }
 
 // onOwnerCreateFunc modify creation condition.
-func onOwnerCreateFunc() func(event.CreateEvent) bool {
+func (r *XGBoostJobReconciler) onOwnerCreateFunc() func(event.CreateEvent) bool {
 	return func(e event.CreateEvent) bool {
 		xgboostJob, ok := e.Object.(*xgboostv1.XGBoostJob)
 		if !ok {
 			return true
 		}
-		xgboostv1.SetDefaults_XGBoostJob(xgboostJob)
-		// TODO: figure out why default funcs are not registered successfully.
-		scheme.Scheme.Default(xgboostJob)
+		r.Scheme.Default(xgboostJob)
 		msg := fmt.Sprintf("xgboostJob %s is created.", e.Object.GetName())
 		logrus.Info(msg)
 

